// @tgrosinger/md-advanced-tables/lib/calc/results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Value = exports.Arity = void 0;
const lodash_1 = require("lodash");
class Arity {
    constructor(rows, columns) {
        this.isRow = () => this.rows > 1 && this.cols === 1;
        this.isColumn = () => this.rows === 1 && this.cols > 1;
        this.isCell = () => this.rows === 1 && this.cols === 1;
        this.rows = rows;
        this.cols = columns;
    }
}
exports.Arity = Arity;
class Value {
    constructor(val) {
        this.get = (row, column) => this.val[row][column];
        this.getAsFloat = (row, column) => {
            const parsed = parseFloat(this.get(row, column));
            return isNaN(parsed) ? 0 : parsed;
        };
        this.getAsInt = (row, column) => {
            const parsed = parseInt(this.get(row, column));
            return isNaN(parsed) ? 0 : parsed;
        };
        /**
         * getArity returns the dimensions of the contained value, in rows and columns
         */
        this.getArity = () => {
            const maxCols = this.val.reduce((max, currentRow) => Math.max(max, currentRow.length), 0);
            return new Arity(this.val.length, maxCols);
        };
        this.toString = () => {
            if (this.getArity().isCell()) {
                return this.get(0, 0);
            }
            return `[${lodash_1.flatten(this.val)
                .map((val) => val.trim())
                .filter((val) => val !== '')
                .join(', ')}]`;
        };
        this.val = val;
    }
}
exports.Value = Value;
