// @tgrosinger/md-advanced-tables/lib/neverthrow/neverthrow.js
"use strict";
/**
 * Retrieved from
 *   https://github.com/supermacro/neverthrow
 *
 * MIT License
 * Copyright (c) 2019 Giorgio Delgado
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Err = exports.Ok = exports.err = exports.ok = void 0;
const ok = (value) => new Ok(value);
exports.ok = ok;
const err = (err) => new Err(err);
exports.err = err;
class Ok {
    constructor(value) {
        this.value = value;
        this.match = (ok, _err) => ok(this.value);
    }
    isOk() {
        return true;
    }
    isErr() {
        return !this.isOk();
    }
    map(f) {
        return exports.ok(f(this.value));
    }
    mapErr(_f) {
        return exports.ok(this.value);
    }
    // add info on how this is really useful for converting a
    // Result<Result<T, E2>, E1>
    // into a Result<T, E2>
    andThen(f) {
        return f(this.value);
    }
    unwrapOr(_v) {
        return this.value;
    }
    _unsafeUnwrap() {
        return this.value;
    }
    _unsafeUnwrapErr() {
        throw new Error('Called `_unsafeUnwrapErr` on an Ok');
    }
}
exports.Ok = Ok;
class Err {
    constructor(error) {
        this.error = error;
        this.match = (_ok, err) => err(this.error);
    }
    isOk() {
        return false;
    }
    isErr() {
        return !this.isOk();
    }
    map(_f) {
        return exports.err(this.error);
    }
    mapErr(f) {
        return exports.err(f(this.error));
    }
    andThen(_f) {
        return exports.err(this.error);
    }
    unwrapOr(v) {
        return v;
    }
    _unsafeUnwrap() {
        throw new Error('Called `_unsafeUnwrap` on an Err');
    }
    _unsafeUnwrapErr() {
        return this.error;
    }
}
exports.Err = Err;
